### PR TITLE
Fixed #24914 -- Added basic permission mixins to contrib.auth

### PR DESCRIPTION
**Pull request by raphaelm**
https://code.djangoproject.com/ticket/24914

An initial version of implementation and tests is in the pull request, documentation is yet to be done. Comments on the code are very welcome.

This proposal provides all the and only the functionality of all three authentication decorators from contrib.auth.

While it is slightly differently implemented, it tries to stay compatible with the widely-used 3rd-party package django-braces in terms of parameter and method names. django-braces does have more functionality than the decorators have, but implementing this functionality in django itself is outside the scope of this ticket.

### CODE REVIEW

**Comment by ewjoachim** [comment]
Saw a few things and told @raphaelm directly, looks all good to me now (except missing documentation ^^)

**Comment by raphaelm** [comment]
I added documentation now. As I'm not a native speaker, I'd love someone to review the language.

I decieded to drop a section in the documentation on class-based view on wrapping as_view with decorators, as there no longer seems to be any decorator in django core where this would make any sense.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by raphaelm** [comment]
Fixed another small documentation problem right now.

**Comment by raphaelm** [comment]
Fixed commit message

**Comment by MarkusH** [comment]
@raphaelm How much of the django-braces stuff did you "copy" or how much is the code / tests inspired by them? Does it make sense to have a acknowledgement notice at the top of the files?

**Comment by raphaelm** [comment]
@MarkusH I did not copy that much, as django-braces has an abstract ``AccessView`` and then derives the three mixins from that, whereas my patch uses UserPassesTestMixin as the base view and the other ones derive from that. I also changed some of smaller implementation details to make it more consistend with the decorators or to be just more readable. The tests are only inspired by the decorator tests from the django codebase.

However, acknowledging them seems nice. Is there a standard way to do this?

File: django/contrib/auth/mixins.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    def get_test_func(self):
+        return getattr(self, "test_func")
+
+    def dispatch(self, request, *args, **kwargs):
+        user_test_result = self.get_test_func()(request.user)
+
+        if not user_test_result:
+            return self.handle_no_permission(request)
+
+        return super(UserPassesTestMixin, self).dispatch(
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: You can join this and a couple of other lines. We have a max line length of 119 chars.

File: django/contrib/auth/mixins.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            request, *args, **kwargs)
+
+    def handle_no_permission(self, request):
+        # In case the 403 handler should be called raise the exception
+        if self.raise_exception:
+            raise PermissionDenied
+        return redirect_to_login(request.get_full_path(),
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: Can you use hanging indents and a trailing comma, please:

```foo(
    1,
    2,
)
```

**Comment by raphaelm** [comment]
@MarkusH I updated the pull request, both of your style comments should be fixed now. Thank you for pointing it out.

File: django/contrib/auth/mixins.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    Mixin for class-based views that checks whether a user has a particular permission
+    enabled, redirecting to the log-in page if necessary.
+    If the raise_exception parameter is given the PermissionDenied exception
+    is raised.
+    """
+    permission_required = None
+
+    def test_func(self, user):
+        if isinstance(self.permission_required, six.string_types):
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: Can you switch to the same logic as in the decorator: `if isinstance(..., (list, tuple)): ...`

Comment (raphaelm) [inline django/contrib/auth/mixins.py]: I can, absolutely, if you think I should. However, I'd rather adjust the decorator's logic to mine: has_perm will take any iterable, therefore restricting it to list and tuple does not seem logical to me. Passing a set can make perfect sense. Therefore, I used the same logic as Django uses in other places where you can pass an iterable or a string (like model._meta.ordering). What do you think?

Am 5. Juni 2015 00:34:33 GMT+01:00, schrieb Markus Holtermann <notifications@github.com>:

Comment (MarkusH) [inline django/contrib/auth/mixins.py]: Valid point. Feel free to change the decorator in a separate commit.

**Comment by MarkusH** [comment]
Regarding the acknowledgement, I'd add something like

```
The authentication mixins UserPassesTestMixin, LoginRequiredMixin and
PermissionRequiredMixin have been inspired by django-braces
<https://github.com/brack3t/django-braces/>
```

**Comment by raphaelm** [comment]
I added the acknowledgement and I'm preparing the logic change for the decorator now, if you'd like to triage that: https://code.djangoproject.com/ticket/24929#ticket

File: django/contrib/auth/mixins.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        return force_text(login_url)
+
+    def get_redirect_field_name(self):
+        """
+        Override this method to customize the redirect_field_name.
+        """
+        if self.redirect_field_name is None:
+            raise ImproperlyConfigured(
+                '{0} has redirect_field_name set to None. Set {0}.redirect_field_name '
+                'to a strng or or override {0}.get_redirect_field_name().'.format(
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: "string"

File: docs/releases/1.9.txt [hunk 5]
```diff
@@ -70,6 +70,11 @@ Minor features
   a deprecation warning in older versions and is no longer supported in
   Django 1.9).
 
+* The module ``django.contrib.auth.mixins`` was added with ``LoginRequiredMixin`,
+  ``PermissionRequiredMixin`` and ``UserPassesTestMixin`` mixins to provide the
+  functionality of the `django.contrib.auth.decorators` shortcuts to class-based
```
Comment (MarkusH) [inline docs/releases/1.9.txt]: double back-tics, please.

File: docs/releases/1.9.txt [hunk 6]
```diff
@@ -70,6 +70,11 @@ Minor features
   a deprecation warning in older versions and is no longer supported in
   Django 1.9).
 
+* The module ``django.contrib.auth.mixins`` was added with ``LoginRequiredMixin`,
```
Comment (MarkusH) [inline docs/releases/1.9.txt]: Please link to the mixins:
```
:class:`~django.contrib.auth.mixins.LoginRequiredMixin`
```

File: docs/topics/auth/default.txt [hunk 7]
```diff
@@ -500,6 +500,40 @@ The login_required decorator
     :func:`django.contrib.admin.views.decorators.staff_member_required`
     decorator a useful alternative to ``login_required()``.
 
+The LoginRequired mixin
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If you are using :doc:`class-based views </topics/class-based-views/index>`, you can
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: "s/If/When" I think.

File: docs/topics/auth/default.txt [hunk 8]
```diff
@@ -500,6 +500,40 @@ The login_required decorator
...
+    for the redirect. Otherwise, :setting:`settings.LOGIN_URL <LOGIN_URL>`
+    will be used. By default, this URL will be called with a query string parameter
+    ``"next"`` containing the URL the user should be redirected to after a successful
+    login. You can change this name by setting the ``redirect_field_name`` parameter
+    on your view. Example::
+
+        from django.contrib.auth.mixins import LoginRequiredMixin
+
+        class MyView(LoginRequiredMixin, View):
+            login_url = '/accounts/login/'
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: Can you change that to something different than settings.LOGIN_URL, e.g. `/login/`

File: docs/topics/auth/default.txt [hunk 9]
```diff
@@ -500,6 +500,40 @@ The login_required decorator
...
+
+.. class:: django.contrib.auth.mixins.LoginRequiredMixin
+
+    .. versionadded:: 1.9
+
+    If a view is using this mixin, all requests by non-authenticated users
+    will be redirected to the login page.
+
+    If you set the ``login_url`` property on your view, this URL will be used
+    for the redirect. Otherwise, :setting:`settings.LOGIN_URL <LOGIN_URL>`
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: use
```
:settings:`~settings.LOGIN_URL`
```
to only display the variable name without the module.

File: docs/topics/auth/default.txt [hunk 10]
```diff
@@ -560,6 +594,32 @@ redirects to the login page::
...
             ...
 
+.. class:: django.contrib.auth.mixins.UserPassesTestMixin
+
+    .. versionadded:: 1.9
+
+    When using :doc:`class-based views </topics/class-based-views/index>`, you can
+    use the ``UserPassesTestMixin`` to do this.
+
+    You should override the ``test_func`` method of the class to provide
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: You have to override ...

File: docs/topics/auth/default.txt [hunk 11]
```diff
@@ -500,6 +500,40 @@ The login_required decorator
...
+
+    If a view is using this mixin, all requests by non-authenticated users
+    will be redirected to the login page.
+
+    If you set the ``login_url`` property on your view, this URL will be used
+    for the redirect. Otherwise, :setting:`settings.LOGIN_URL <LOGIN_URL>`
+    will be used. By default, this URL will be called with a query string parameter
+    ``"next"`` containing the URL the user should be redirected to after a successful
+    login. You can change this name by setting the ``redirect_field_name`` parameter
+    on your view. Example::
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: I'd strip the "Example" and instead use "on your view::"

File: docs/topics/auth/default.txt [hunk 12]
```diff
@@ -560,6 +594,32 @@ redirects to the login page::
...
 
+.. class:: django.contrib.auth.mixins.UserPassesTestMixin
+
+    .. versionadded:: 1.9
+
+    When using :doc:`class-based views </topics/class-based-views/index>`, you can
+    use the ``UserPassesTestMixin`` to do this.
+
+    You should override the ``test_func`` method of the class to provide
+    the actual test that is performed. Example::
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: Chop Example, as above.

File: docs/topics/auth/default.txt [hunk 13]
```diff
@@ -605,10 +665,30 @@ Applying permissions to generic views
...
+</ref/class-based-views/index>`, you can use the following mixin:
+
+.. class:: django.contrib.auth.mixins.PermissionRequiredMixin
+
+    .. versionadded:: 1.9
+
+    This mixin, just like the ``permisison_required``
+    decorator, checks whether the user accessing a view has a particular
+    permission. You should specify the permission (or a list or permissions)
+    using the ``permission_required`` parameter. Example::
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: Chop Example, as above

**Comment by MarkusH** [comment]
There's a bunch of docs duplication with regards to `login_url`, `raise_exception` and `redirect_field_name`. I'd document them on the `UserPassesTestMixin` and reference their documentation from `LoginRequiredMixin` and `PermissionRequiredMixin` instead.

File: django/contrib/auth/mixins.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+from django.utils.encoding import force_text
+
+
+class UserPassesTestMixin(object):
+    """
+    Mixin for class-based views that checks that the user passes the given by
+    the class method `test_func`.
+    """
+    login_url = None
+    redirect_field_name = REDIRECT_FIELD_NAME
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: Can you sort those attributes please

File: django/contrib/auth/mixins.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    def test_func(self, user):
+        raise NotImplementedError(
+            '{0} is missing implementation of the '
+            'test_func method.'.format(self.__class__.__name__)
+        )
+
+    def get_test_func(self):
+        return getattr(self, "test_func")
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: I'd keep `dispatch()` the first method on the view.

File: tests/auth_tests/test_mixins.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+from django.test import RequestFactory, TestCase, override_settings
+from django.views.generic import View
+
+from .test_views import AuthViewsTestCase
+
+
+@override_settings(ROOT_URLCONF='auth_tests.urls')
+class LoginRequiredMixinTestCase(AuthViewsTestCase):
+    """
+    Tests the LoginRequired mixin
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: I don't think this docstring is required.

File: tests/auth_tests/test_mixins.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+    """
+    def setUp(self):
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: There's a class method called `setUpTestData()`. Please use that instead.

File: tests/auth_tests/test_mixins.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        self.user = models.User.objects.create(username='joe', password='qwerty')
+        self.factory = RequestFactory()
+
+    def testLoginRequired(self, view=None, login_url=None):
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: use snake case, please: `test_login_required()`

**Comment by MarkusH** [comment]
I'm missing tests for the `UserPassesTestMixin`.

**Comment by raphaelm** [comment]
I updated the pull request and fixed/changed everything you remarked, except those two:

> There's a bunch of docs duplication with regards to login_url, raise_exception and redirect_field_name. I'd document them on the UserPassesTestMixin and reference their documentation from LoginRequiredMixin and PermissionRequiredMixin instead.

I am unsure whether this is the best solution. I feel that UserPassesTestMixin is the least important one of those and most users will only be interested in LoginRequiredMixin and maybe PermissionRequiredMixin. Therefore, I think it would be more confusing then helpful.

> I'm missing tests for the UserPassesTestMixin.

Me too, but how would you test this properly? UserPassesTest only makes sense to use if you override test_func and test that subclass -- however, this is exactly what the tests for LoginRequiredMixin and PermissionRequiredMixin do. All aspects of UserPassesTest are tested, but I do not see a nice way to do it directly.

(There is no test for the user_passes_test decorator, I guess for the same reason. Also, most mistakes you pointed out in my code in your last comments, are present in the decorators as well.)

**Comment by raphaelm** [comment]
Fixed spelling now, but the docs build still fails because of broken links. I do not see the mistake, do you see it?

**Comment by raphaelm** [comment]
Okay, I found the problem, documentation builds now.

**Comment by timgraham** [comment]
FWIW, I'd give acknowledgement in AUTHORS and/or the commit message. That seems more in line with the 4 bullet here: https://docs.djangoproject.com/en/1.8/internals/contributing/writing-code/coding-style/#miscellaneous

**Comment by MarkusH** [comment]
Regarding documentation duplication: The arguments are implemented on the `UserPassesTestMixin`, which I think implies that they are documented there.

File: tests/auth_tests/test_mixins.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+                def get(self, request, *args, **kwargs):
+                    return HttpResponse()
+
+            view = AView.as_view()
+
+        request = self.factory.get('/rand')
+        request.user = AnonymousUser()
+        response = view(request)
+        self.assertEqual(response.status_code, 302)
+        self.assertIn(login_url, response.url)
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: Isn't there something like `assertRedirects()`?

**Comment by MarkusH** [comment]
Regarding the missing tests: I'd create 2 mixins in the tests that have `def test_func(self, user): return True` (`False` respectively) and check all the combinations of `login_url`, `raise_exception`, and `redirect_field_name`. The `LoginRequiredMixin` and `PermissionRequiredMixin` could then just check one positive and one negative use case.

**Comment by raphaelm** [comment]
Okay, I'll look into this during the next days.

Am 6. Juni 2015 16:12:32 MESZ, schrieb Markus Holtermann <notifications@github.com>:

File: tests/auth_tests/test_mixins.py (hunk 19 above)
Comment (raphaelm) [inline tests/auth_tests/test_mixins.py]: Yes, but apparently it only works with the test client, not with RequestFactory. I'll change the assertIn to an equality check, but I don't think it is worth creating a bunch of URLs for the test views in the URL config just to test these mixins.

**Comment by raphaelm** [comment]
> FWIW, I'd give acknowledgement in AUTHORS and/or the commit message.

I'd go for the commit message than as its only a few lines coming from there.

I updated the documentation and tests according to all your other comments.

**Comment by raphaelm** [comment]
Just updated the PR again, now rebased to current ``master`` without any merge conflicts.

File: tests/auth_tests/test_mixins.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+    def test_user_passes(self):
+        view = UserTestAlwaysTrueView.as_view()
+        request = self.factory.get('/rand')
+        request.user = AnonymousUser()
+        response = view(request)
+        self.assertEqual(response.status_code, 200)
+
+
+class LoginRequiredMixinTestCase(AuthViewsTestCase):
+    @classmethod
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: One empty line above, please.

File: docs/topics/auth/default.txt [hunk 21]
```diff
@@ -560,6 +595,47 @@ redirects to the login page::
...
             ...
 
+
+.. currentmodule:: django.contrib.auth.mixins
+
+.. class:: UserPassesTestMixin
+
+    .. versionadded:: 1.9
+
+    When using :doc:`class-based views </topics/class-based-views/index>`, you can
```
Comment (MarkusH) [inline docs/topics/auth/default.txt]: Can you wrap the documentation at 79 chars, please.

File: tests/auth_tests/test_mixins.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual('/accounts/login/?next=/rand', response.url)
+        request = self.factory.get('/rand')
+        request.user = self.user
+        response = view(request)
+        self.assertEqual(response.status_code, 200)
+
+
+class PermissionsRequiredMixinTest(TestCase):
+    @classmethod
```
Comment (MarkusH) [inline tests/auth_tests/test_mixins.py]: As above.

File: django/contrib/auth/mixins.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+    redirecting to the log-in page if necessary.
+    """
+
+    def test_func(self, user):
+        return user.is_authenticated()
+
+
+class PermissionRequiredMixin(UserPassesTestMixin):
+    """
+    Mixin for class-based views that checks whether a user has a particular permission
```
Comment (MarkusH) [inline django/contrib/auth/mixins.py]: Wrap at 79 chars, please.

**Comment by raphaelm** [comment]
Style issues fixed, PR updated.

**Comment by MarkusH** [comment]
Replaced by #4836

**Pull request closed without merging**